### PR TITLE
Seed the experience index and LRS with synthetic data

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

We need to load experiences in the experience index for development / demo purpose.

## Proposal

Given an input statements archive, we seed the experience index with courses and activities.

```Usage: python seed_experience_index.py [ARCHIVE]```

If no archive file is given as input, we parse statements from the standard input.

### CODE REVIEW

File: bin/seed_experience_index.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    """
+    stream = sys.stdin
+    if len(sys.argv) == 2:
+        logger.info("Reading statements from archive: %s", sys.argv[1])
+        stream = Path(sys.argv[1]).open()
+    else:
+        logger.info("Reading statements from stdin")
+
+    course_activities = parse_statements(stream)
+    seed_experience_index(course_activities)
```
Comment (jmaupetit) [inline bin/seed_experience_index.py]:
```suggestion
    seed_experience_index(parse_statements(stream))
```

File: bin/seed_experience_index.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    # Add the new experiences to the Experience Index
+    try:
+        session.add_all(experiences)
+        session.commit()
+    except SQLAlchemyError:
+        logging.debug("Could not seed the Experience Index. Exception:", exc_info=True)
+        session.rollback()
+    finally:
+        session.close()
```
Comment (jmaupetit) [inline bin/seed_experience_index.py]: Not if we create relations.

File: bin/seed_experience_index.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        logger.debug("Action: %s", action)
+
+        yield (course, action)
+
+
+def seed_experience_index(course_actions: Generator[Tuple[Course, Action], None, None]):
+    """Seed the Experience Index with mocked data.
+
+    The mocked data have to align with the Learning Record Store (LRS).
+    Note: If there are updates to the LRS fixtures, manually update the IRI accordingly.
```
Comment (jmaupetit) [inline bin/seed_experience_index.py]: To update.

File: bin/seed_experience_index.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        course_action_iris += [(course.iri, action.iri)]
+
+    logger.info(
+        "Will create %d experiences (%d courses + %d actions)",
+        len(experiences),
+        len(courses),
+        len(actions),
+    )
+    logger.debug("Courses IRIs: %s", courses)
+    logger.debug("actions IRIs: %s", actions)
```
Comment (jmaupetit) [inline bin/seed_experience_index.py]:
```suggestion
    logger.debug("Actions IRIs: %s", actions)
```

File: bin/seed_experience_index.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
+"""Seed the experience index"""
```
Comment (quitterie-lcs) [inline bin/seed_experience_index.py]:
```suggestion
"""Seed the experience index."""
```

File: bin/seed_experience_index.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    course_action_relations = pd.DataFrame(columns=["course_id", "action_id"])
+    chunk_size = 1000
+    for idx in range(0, course_action_iris.shape[0], chunk_size):
+        logger.info(
+            "Create course/action IDs dataframe for chunk %d (size: %d)",
+            idx,
+            chunk_size,
+        )
+        # Add course IDs from database given course IRIs
+        tmp_relations = pd.merge(
```
Comment (quitterie-lcs) [inline bin/seed_experience_index.py:215]: I don't get why you have to create a temporary dataframe. I think you can concat directly based on a condition on IRIS, no ?

Comment (jmaupetit) [reply bin/seed_experience_index.py:215]: As merging two big dataframes could be time-consuming, splitting those merge events in smaller chunks is more efficient.

Comment (quitterie-lcs) [reply bin/seed_experience_index.py:215]: alright

**Review by quitterie-lcs — APPROVED** [review]

**Pull request merged**